web_src/js/utils.js

```diff
@@ -66,7 +66,7 @@ export function strSubMatch(full, sub) {
   let i = 0, j = 0;
   while (i < sub.length && j < full.length) {
     while (j < full.length) {
-      if (sub[i] === full[j]) {
+      if (typeof sub[i] !== 'undefined' && typeof full[j] !== 'undefined' && sub[i].toLowerCase() === full[j].toLowerCase()) {
```

Comment:
Why are the type checks necessary?

Reply:
Because I noticed that `sub[i]` was undefined under some circumstances (and a call to `toLowerCase()` would thus error out.

I am guessing this could be due to the `i++` instruction on line 73, since nothing checks whether `i` is still within `sub`'s length or not.

I preferred staying on the safe side and do a type check on both variables instead of refactoring the loop and potentially making some edge case different.

Checking whether `i < sub.length` would effectively make the while loop on line 68 identical to the one on line 67.. which looks odd? Again, I preferred not refactoring the whole algorithm for this edge case, and instead type-checking this.

Would you like me to handle this differently?

Reply:
Should just write `sub[i]?.toLowerCase() === full[j]?.toLowerCase()` if typecheck is really necessary, depends on how these arrays are generated. If they are guaranteed to be strings, the `?` is unnecessary.

Reply:
`sub` and `full` are strings actually! the issue is that `i` can get too big, and in all honesty, I'm slightly confused to why a double-loop is necessary here at all.

Reply:
> Checking whether i < sub.length would effectively make the while loop on line 68 identical to the one on line 67.. which looks odd? Again, I preferred not refactoring the whole algorithm for this edge case, and instead type-checking this.

Yes, the inner loop looks strange. If you remove it, one of the tests fails but maybe the test is wrong...

Reply:
works:
```
  while (i < sub.length && j < full.length) {
    while (j < full.length) {
```
does not work:
```
  while (i < sub.length && j < full.length) {
```
works and should not need type checks:
```
  while (i < sub.length && j < full.length) {
    while (i < sub.length && j < full.length) {
```
I will have a look with the debugger why the second one does not work.

Reply:
The tests are correct! I think the current algorithm needs to loop over the full string multiple times (which happens when the inner loop is broken from line 78).

FWIW I just committed silverwind's suggestion (I completely forgot this was a thing in JS, my bad 😅)

Reply:
Yes, the break needs to get removed. Then there should be no typecheck needed.

Reply:
```js
  while (i < sub.length && j < full.length) {
    if (sub[i].toLowerCase() === full[j].toLowerCase()) {
      if (res.length % 2 !== 0) res.push('');
        res[res.length - 1] += full[j];
        j++;
        i++;
      } else {
        if (res.length % 2 === 0) res.push('');
        res[res.length - 1] += full[j];
        j++;
      }
    }
  }
```

Reply:
Done! I wonder why the code was like this in the first place...